framework/src/araui/libs/upng.c

```diff
@@ -0,0 +1,1299 @@
+		if ((bp >> 3) >= insize) {
+			SET_ERROR(upng, UPNG_EMALFORMED);
+			return upng->error;
+		}
+
+		/* read block control bits */
+		done = read_bit(&bp, &in[inpos]);
+		btype = read_bit(&bp, &in[inpos]) | (read_bit(&bp, &in[inpos]) << 1);
+
+		/* process control type appropriateyly */
+		if (btype == 3) {
+			SET_ERROR(upng, UPNG_EMALFORMED);
+			return upng->error;
+		} else if (btype == 0) {
+			inflate_uncompressed(upng, out, outsize, &in[inpos], &bp, &pos, insize);	/*no compression */
+		} else {
+			inflate_huffman(upng, out, outsize, &in[inpos], &bp, &pos, insize, btype);	/*compression, btype 01 or 10 */
+		}
+
+		/* stop if an error has occured */
+		if (upng->error != UPNG_EOK) {
+			return upng->error;
+		}
+	}
+
+	return upng->error;
+}
+
+static upng_error uz_inflate(upng_t* upng, unsigned char *out, unsigned long outsize, const unsigned char *in, unsigned long insize)
+{
+	/* we require two bytes for the zlib data header */
+	if (insize < 2) {
+		SET_ERROR(upng, UPNG_EMALFORMED);
+		return upng->error;
+	}
+
+	/* 256 * in[0] + in[1] must be a multiple of 31, the FCHECK value is supposed to be made that way */
+	if ((in[0] * 256 + in[1]) % 31 != 0) {
+		SET_ERROR(upng, UPNG_EMALFORMED);
+		return upng->error;
+	}
+
+	/*error: only compression method 8: inflate with sliding window of 32k is supported by the PNG spec */
+	if ((in[0] & 15) != 8 || ((in[0] >> 4) & 15) > 7) {
+		SET_ERROR(upng, UPNG_EMALFORMED);
+		return upng->error;
+	}
+
+	/* the specification of PNG says about the zlib stream: "The additional flags shall not specify a preset dictionary." */
+	if (((in[1] >> 5) & 1) != 0) {
+		SET_ERROR(upng, UPNG_EMALFORMED);
+		return upng->error;
+	}
+
+	/* create output buffer */
+	uz_inflate_data(upng, out, outsize, in, insize, 2);
+
+	return upng->error;
+}
+
+/*Paeth predicter, used by PNG filter type 4*/
+static int paeth_predictor(int a, int b, int c)
+{
+	int p = a + b - c;
+	int pa = p > a ? p - a : a - p;
+	int pb = p > b ? p - b : b - p;
+	int pc = p > c ? p - c : c - p;
+
+	if (pa <= pb && pa <= pc)
+		return a;
+	else if (pb <= pc)
+		return b;
+	else
+		return c;
+}
+
+static void unfilter_scanline(upng_t* upng, unsigned char *recon, const unsigned char *scanline, const unsigned char *precon, unsigned long bytewidth, unsigned char filterType, unsigned long length)
+{
+	/*
+	   For PNG filter method 0
+	   unfilter a PNG image scanline by scanline. when the pixels are smaller than 1 byte, the filter works byte per byte (bytewidth = 1)
+	   precon is the previous unfiltered scanline, recon the result, scanline the current one
+	   the incoming scanlines do NOT include the filtertype byte, that one is given in the parameter filterType instead
+	   recon and scanline MAY be the same memory address! precon must be disjoint.
+	 */
+
+	unsigned long i;
+	switch (filterType) {
+	case 0:
+		for (i = 0; i < length; i++)
+			recon[i] = scanline[i];
+		break;
+	case 1:
+		for (i = 0; i < bytewidth; i++)
```

Comment:
How about below?
```
for (i =0; i < length; i++) {
   recon[i] = scanline[i];
   if (i > bytewidth) {
     recon[i] += recon[i - bytewidth];
   }
}

Reply:
This is an open source lib code. For maintainability, I'm going to remain this.